Sources/PubNub/Helpers/Crypto/CryptorModule.swift

```diff
@@ -0,0 +1,320 @@
+  }
+  
+  /// Encrypts the given `Data` object
+  /// 
+  /// - Parameters:
+  ///   - data: Data to encrypt
+  /// - Returns: A success, storing encrypted `Data` if operation succeeds. Otherwise, a failure storing `PubNubError` is returned
+  public func encrypt(data: Data) -> Result<Data, PubNubError> {
+    guard !data.isEmpty else {
+      return .failure(PubNubError(
+        .encryptionError,
+        additional: ["Cannot encrypt empty Data"])
+      )
+    }
+    return defaultCryptor.encrypt(data: data).map {
+      if defaultCryptor.id == LegacyCryptor.ID {
+        return $0.data
+      } else {
+        return CryptorHeader.v1(
+          cryptorId: defaultCryptor.id,
+          data: $0.metadata
+        ).asData() + $0.data
+      }
+    }.mapError {
+      PubNubError(.encryptionError, underlying: $0)
+    }
+  }
+  
+  /// Decrypts the given `Data` object
+  ///
+  /// - Parameters:
+  ///   - data: Data to encrypt
+  /// - Returns: A success, storing decrypted `Data` if operation succeeds. Otherwise, a failure storing `PubNubError` is returned
+  public func decrypt(data: Data) -> Result<Data, PubNubError> {
+    guard !data.isEmpty else {
+      return .failure(PubNubError(
+        .decryptionError,
+        additional: ["Cannot decrypt empty Data"])
+      )
+    }
+    do {
+      let header = try CryptorHeader.from(data: data)
+      
+      guard let cryptor = cryptor(matching: header) else {
+        return .failure(PubNubError(
+          .unknownCryptorError,
+          additional: [
+            "Could not find matching Cryptor for \(header.cryptorId()) while decrypting Data. " +
+            "Ensure the corresponding instance is registered in \(String(describing: Self.self))"
+          ]
+        ))
+      }
+      return cryptor.decrypt(
+        data: EncryptedData(
+          metadata: header.metadataIfAny(),
+          data: data.subdata(in: header.length()..<data.count)
+        )
+      )
+      .flatMap {
+        if $0.isEmpty {
+          return .failure(PubNubError(
+            .decryptionError,
+            additional: ["Decrypting resulted with empty Data"])
+          )
+        }
+        return .success($0)
+      }
+      .mapError {
+        PubNubError(.decryptionError, underlying: $0)
+      }
+    } catch let error as PubNubError {
+      return .failure(error)
+    } catch {
+      return .failure(PubNubError(
+        .decryptionError,
+        underlying: error,
+        additional: ["Cannot decrypt InputStream"])
+      )
+    }
+  }
+  
+  /// Encrypts the given `InputStream` object
+  ///
+  /// - Parameters:
+  ///   - stream: Stream to encrypt
+  ///   - contentLength: Content length of encoded stream
+  /// - Returns: A success, storing an ``EncryptedStreamResult`` value if operation succeeds. Otherwise, a failure storing `PubNubError` is returned
+  public func encrypt(stream: InputStream, contentLength: Int) -> Result<EncryptedStreamResult, PubNubError> {
```

Comment:
Do we really need `EncryptedStreamResult` in response? Can we transparently use `InputStream` both for the source and resulting streams?
I think in tests, I've noticed that we would like to ask user to create `EncryptedStreamResult` to make encryption. What do you think if user will just provide regular `InputStream` and the size of data in it and internally in `CryptoModule.encrypt` method we will do the rest and wrap it into `EncryptedStreamResult`?

Reply:
Looks possible. We could return a tuple (under some `typealias` or not) in the response that gathers `InputStream` and its content length. The second part (content length) is needed in order to fill `Content-Length` HTTP header field when uploading a File. This is [how](https://github.com/pubnub/swift/blob/feat/cryptors/Sources/PubNub/Extensions/URLRequest%2BPubNub.swift#L80) I made it in my PR.